File: CPP/ByRound/1861/1861_B_Two_Binary_Strings.cpp
// Problem: CF 1861 B - Two Binary Strings
// https://codeforces.com/contest/1861/problem/B

/*
B. Two Binary Strings

Algorithm/Technique: Direct comparison of adjacent character pairs in both strings.

Time Complexity: O(n) where n is the length of the string (for each test case).
Space Complexity: O(1) (only using a few variables, no extra space proportional to input size).

The problem allows merging equal characters in a string to make it uniform between them.
We can only perform operations that merge segments of identical characters.
The key insight is that we just need to check whether there exists a pair of adjacent characters
in both strings such that one is '0' and the other is '1', or vice versa, because:
- If at any point both strings have same character pattern, then all transitions can be made equal.
- So, if we find even one place where both strings match in '0' followed by '1', or '1' followed by '0',
  it's possible to make strings equal with operations.

However, the current implementation checks for a specific case:
- A transition from '0' to '1' in both strings at some index p (i.e., a[p] = '1' && b[p] = '1' && a[p-1] = '0' && b[p-1] = '0')
  This condition ensures there's an opportunity to align parts of the strings.

If such a pair is found, it means we can possibly merge or adjust parts to make strings equal.
*/

#include <iostream>
#include <string>

int main() {
  std::ios_base::sync_with_stdio(false); // Optimize I/O speed
  long t;
  std::cin >> t;
  
  while (t--) {
    std::string a, b;
    std::cin >> a >> b;
    
    bool possible(false); // Flag to determine if it's possible to make strings equal
    
    // Loop through all positions except the first and last ones
    for (long p = 1; !possible && p < a.size(); p++) {
      // Check if both strings have '1' at position p, and '0' at position p-1
      // This condition implies there's a potential to align both strings
      if (a[p] == '1' && b[p] == '1' && a[p - 1] == '0' && b[p - 1] == '0') {
        possible = true;
      }
    }
    
    // Output result based on flag
    std::cout << (possible ? "YES" : "NO") << std::endl;
  }
}



// https://github.com/VaHiX/codeForces/

File: CSharp/ByTier/B/1861_B_Two_Binary_Strings.cs
// Problem: CF 1861 B - Two Binary Strings
// https://codeforces.com/contest/1861/problem/B

/*
Algorithm: Two Binary Strings
Techniques: Simulation with StringBuilder, Greedy approach

Time Complexity: O(n^2) where n is the length of the strings. In the worst case, for each position we might traverse the entire string.
Space Complexity: O(n) for storing the strings using StringBuilder.

The problem involves making two binary strings equal using a specific operation:
- Choose a string and two equal characters at positions l and r (l < r) to turn all characters between them into that character.
- We must determine if we can make both strings completely equal.

The solution uses a greedy simulation approach:
- Start at index 1 (since index 0 is always '0' and index n-1 is always '1').
- At each step, check if both strings have '1' at current position:
  - If yes, it's possible to align, so answer is YES.
- If only one has '1':
  - Remove all consecutive '1's from that string starting at current index.
  - If we reach end of string without finding a '0' to continue, return NO.
- Continue until the end of string is reached.

*/
using System;
using System.Text;
public class TwoBinaryStrings {
  public static void Main() {
    short t = short.Parse(Console.ReadLine());
    for (short i = 0; i < t; i++) {
      string a = Console.ReadLine();
      string b = Console.ReadLine();
      short len = Convert.ToInt16(a.Length);
      StringBuilder sba = new StringBuilder(a);
      StringBuilder sbb = new StringBuilder(b);
      short j = 1;
      bool solved = false;
      bool ans = false;
      while (!solved) {
        // If both strings have '1' at current position, we can make the strings align
        if (sba[j] == '1' && sbb[j] == '1') {
          ans = true;
          solved = true;
        } else if (sba[j] == '1' || sbb[j] == '1') {
          // If only one string has '1', remove all consecutive '1's from that string
          short k = RemoveOnes((sba[j] == '1' ? sba : sbb), j);
          if (k == len) {
            // If we reached the end of string without finding a '0', it's impossible
            ans = false;
            solved = true;
          }
        }
        j++;
      }
      Console.WriteLine(ans ? "YES" : "NO");
    }
  }
  // Helper function to remove consecutive '1's starting from index i
  public static short RemoveOnes(StringBuilder sb, short i) {
    while (i < sb.Length && sb[i] == '1') {
      sb[i] = '0';
      i++;
    }
    return i;
  }
}


// https://github.com/VaHiX/CodeForces/